compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs

```diff
@@ -630,14 +630,32 @@ impl CStore {
         }
     }
 
-    pub(crate) fn update_extern_crate(&mut self, cnum: CrateNum, extern_crate: ExternCrate) {
+    /// Track how an extern crate has been loaded
+    ///
+    /// * the `name` is for [`Self::set_resolved_extern_crate_name`]
+    /// * `extern_crate` is for diagnostics
+    pub(crate) fn update_extern_crate(
+        &mut self,
+        cnum: CrateNum,
+        name: Symbol,
+        extern_crate: ExternCrate,
+    ) {
+        debug_assert_eq!(
+            extern_crate.dependency_of, LOCAL_CRATE,
+            "this function should not be called on transitive dependencies"
+        );
+        self.set_resolved_extern_crate_name(name, cnum);
```

Comment:
I think it would be fine to do `if extern_crate.dependency_of == LOCAL_CRATE { set_resolved_extern_crate_name }` instead of splitting `update_extern_crate` into two function.
I doubt it will cause any perf regressions, but we can run the benchmarks.

Reply:
I've done this for clarity, not performance. Updating local crate imports and updating transitive dependencies are two different tasks.

Logically, the caller should call both:

```
store.set_resolved_extern_crate_name(name);
store.update_extern_crate(deps);
```

but it would be a bit fragile, since the caller could do one thing but forget the other, which is why I have one entrypoint for the local crate, but the transitive deps update separately.

Reply:
@petrochenkov I've split it into 117447f7e8aa01c09a234aaa74265cdfec3d5777 and dfdf5b80e48cb51d24014fff00ab33d6cf5c6758 to demonstrate the difference